rotkehlchen/chain/ethereum/modules/balancer/decoder.py

```diff
@@ -0,0 +1,291 @@
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+class BalancerDecoder(DecoderInterface):
+
+    def __init__(
+            self,
+            evm_inquirer: 'EvmNodeInquirer',
+            base_tools: 'BaseDecoderTools',
+            msg_aggregator: 'MessagesAggregator',
+    ) -> None:
+        super().__init__(
+            evm_inquirer=evm_inquirer,
+            base_tools=base_tools,
+            msg_aggregator=msg_aggregator,
+        )
+        self.base_tools = base_tools
+        self.eth = A_ETH.resolve_to_crypto_asset()
+        self.weth = A_WETH.resolve_to_evm_token()
+
+    def _decode_v1_pool_event(self, all_logs: list[EvmTxReceiptLog]) -> Optional[list[ActionItem]]:
+        """Read the list of logs in search for a Balancer V1 event and return the information
+        needed to decode the transfers made in the transaction to/from the ds proxy
+        """
+        # The transfer event appears after the debt generation event, so we need to transform it
+        target_logs = []
+        for tx_log in all_logs:
+            if tx_log.topics[0] == JOIN_V1 or tx_log.topics[0] == EXIT_V1:
+                target_logs.append(tx_log)
+
+        if len(target_logs) == 0:
+            return None
+
+        action_items = []
+        for target_log in target_logs:
+            token_address = hex_or_bytes_to_address(target_log.topics[2])
+            amount = hex_or_bytes_to_int(target_log.data[0:32])
+            if target_log.topics[0] == JOIN_V1:
+                balancer_event_type = BalancerV1EventTypes.JOIN
+                from_event_type = HistoryEventType.SPEND
+                ds_address = hex_or_bytes_to_address(target_log.topics[1])
+
+            else:
+                balancer_event_type = BalancerV1EventTypes.EXIT
+                from_event_type = HistoryEventType.RECEIVE
+                ds_address = target_log.address
+
+            action_item = ActionItem(
+                action='transform',
+                sequence_index=target_log.log_index,
+                from_event_type=from_event_type,
+                from_event_subtype=HistoryEventSubType.NONE,
+                asset=self.eth,
+                amount=ZERO,
+                to_event_type=None,
+                to_event_subtype=None,
+                to_counterparty=CPT_BALANCER_V1,
+                to_notes=None,
+                extra_data={
+                    'ds_address': ds_address,
+                    'token_address': token_address,
+                    'amount': amount,
+                    'type': balancer_event_type,
+                },
+            )
+            action_items.append(action_item)
+
+        return action_items
+
+    def _decode_swap_creation(
+            self,
+            tx_log: EvmTxReceiptLog,
+            decoded_events: list[HistoryBaseEntry],
+            action_items: list[ActionItem],  # pylint: disable=unused-argument
+    ) -> tuple[Optional[HistoryBaseEntry], list[ActionItem]]:
```

Comment:
Can you write a docstring here xplaining what you do here and how the logic works? The following comment probably needs to go in the docstring.

Also explain what you do more. For example later I see

```python
if len(action_items) == 0 and from_token == self.weth.evm_address:
```

and I don't understand why/what that is. You need to explain code or write self-explainable code (where possible)

Reply:
I have added information to the docstring and in comments to clarify